src/transformers/integrations/awq.py

```diff
@@ -102,3 +120,145 @@ def replace_with_awq_linear(
+    """
+    # Always default to `quantization_config.fusing_mapping`
+    if quantization_config.fusing_mapping is not None:
+        current_fused_mapping = quantization_config.fusing_mapping
+        current_fused_mapping["max_seq_len"] = quantization_config.fuse_max_seq_len
+    elif model.config.model_type in AWQ_FUSED_MAPPINGS:
+        current_fused_mapping = AWQ_FUSED_MAPPINGS[model.config.model_type]
+
+        # Handle hidden_size, num_attention_heads, num_key_value_heads on our own.
+        hidden_size = model.config.hidden_size
+        num_attention_heads = model.config.num_attention_heads
+        num_key_value_heads = getattr(model.config, "num_key_value_heads", num_attention_heads)
+
+        # Fill `current_fused_mapping` with the expected values
+        current_fused_mapping["hidden_size"] = hidden_size
+        current_fused_mapping["num_attention_heads"] = num_attention_heads
+        current_fused_mapping["num_key_value_heads"] = num_key_value_heads
+        current_fused_mapping["max_seq_len"] = quantization_config.fuse_max_seq_len
+    else:
+        raise ValueError(
+            "Fusing mapping not found either on the quantization config or the supported `AWQ_FUSED_MAPPINGS`. Please pass a `fused_mapping` argument"
+            " in the `quantization_config` or raise an issue on transformers https://github.com/huggingface/transformers to add its support."
+        )
+    return current_fused_mapping
+
+
+def fuse_awq_modules(model, quantization_config):
+    """
+    Optionally fuse some modules in the model to speedup inference.
+
+    Args:
+        model (`~PreTrainedModel`):
+            The model to fuse - note this model should have been converted into AWQ format beforehand.
+        quantization_config (`~transformers.quantization_config.AWQConfig`):
+            The quantization configuration to use.
+    """
+    backend = quantization_config.backend
+    fusing_mapping = get_fusing_mapping(model, quantization_config)
+
+    if backend == AwqBackendPackingMethod.AUTOAWQ:
+        from awq.modules.fused.attn import QuantAttentionFused
+        from awq.modules.fused.mlp import QuantFusedMLP
+        from awq.modules.fused.norm import FasterTransformerRMSNorm
+        from awq.modules.linear import WQLinear_GEMM, WQLinear_GEMV
+    else:
+        raise ValueError("Fusing is only supported for the AutoAWQ backend")
+
+    for name, module in model.named_modules():
+        # Replace layer norms
+        for module_name in fusing_mapping["layernorm"]:
+            if hasattr(module, module_name):
+                old_module = getattr(module, module_name)
+                module._modules[module_name] = FasterTransformerRMSNorm(
+                    old_module.weight,
+                    old_module.variance_epsilon,
+                ).to(old_module.weight.device)
+                del old_module
+        # Replace MLP layers
+        if hasattr(module, fusing_mapping["mlp"][0]):
```

Comment:
Can we guarantee `fusing_mapping` has an `"mlp"` key? and that it has at least one value?

Reply:
I propose to do a more extensive check in b187c07 and I changed the logic there a bit to make sure these methods will do nothing in case we put an empty array in these fields